### DISCUSSION TITLE
Converting LaunchIndex to PixelIdx

### DISCUSSION THREAD

**Discussion by asprasan** [owl-project/owl#175] [Q&A] [answered]
In the OWL samples, I see two different ways of converting the LaunchIndex to the corresponding pixelIdx, which are:
```cpp
const vec2i pixelID = owl::getLaunchIndex();
const int pixelIdx = pixelID.x+self.fbSize.x*(self.fbSize.y-1-pixelID.y);
```
and 
```cpp
const vec2i pixelID = owl::getLaunchIndex();
const int pixelIdx = pixelID.x+self.fbSize.x*(pixelID.y);
```

I understand that the `getLaunchIndex()` routine outputs a pixelID where (0,0) is in the top-left corner. With that logic, `const int pixelIdx = pixelID.x+self.fbSize.x*(pixelID.y);` should be the correct way to convert the launch-index to the pixel-index. 
Have I understood this correctly?
If yes, when do we need to use `const int pixelIdx = pixelID.x+self.fbSize.x*(self.fbSize.y-1-pixelID.y);`?

**Comment by szellmann** [comment, marked as answer]
One sample's pixel/window coordinate system seems to have its origin at the bottom, while the other origin is at the top.

OWL doesn't have a concept of what a pixel coordinate system is, so it's totally up to the user.

The viewer however does have such a concept, and as it uses OpenGL, its window coordinate system's origin is bottom/left.

My guess is that one of the samples is interactive, and the other one is a cmdline sample, and the cmdline samples assume that the coordinate system origin is different, because they use SBT to write out images.

Personally - and if that is actually the reason - I would prefer that all the samples use the same origin, and then use `stbi_flip_vertically_on_write(1)` to make SBT flip the image. But yeah, totally up to the user, OWL doesn't even define what a pixel or a frame buffer is.

**Reply by ingowald** [reply]
Ooooh, interesting. I didn't even know that STB can mirror on write - if it can do this then I'd indeed vote for using that procedure as "intended use" (though the answer itself is still right: OWL and optix themselves do not even have a concept of up or down, it's entirely the app's or viewer's business how they interprets that).

If anybody wants to go ahead and change any of the existing samples to follow that convention, i'd be happy to merge that.

**Reply by asprasan** [reply]
Okay. It makes sense to me that OWL and Optix themselves do not have a concept of up/down. But I found both the versions of procedures to convert from launchIndex in `cmdline` samples.

There was no use of OpenGL viewer or anything. So, it's making me curious why `pixelIdx = pixelID.x+self.fbSize.x*(self.fbSize.y-1-pixelID.y);` was used in `s05-rtow` cmdline sample?
